tests/test_tensor/test_optimizers.py

```diff
@@ -214,3 +252,33 @@ def test_multiloss(backend, executor):
 
     if executor is not None:
         executor.shutdown()
+
+
+@pytest.mark.parametrize('backend', [jax_case, autograd_case,
+                                     tensorflow_case])
+def test_shared_tags(tagged_qaoa_tn,backend):
+    n,depth,psi0 = tagged_qaoa_tn
+
+    H = qu.ham_heis(n,j=(0.,0.,-1.),b=(1.,0.,0.),cyclic=True,)
+    gs = qu.groundstate(H)
+    T_gs = qtn.Dense1D(gs)
+
+    def loss(psi, target):
+        f = psi.H & target
+        f.rank_simplify_()
+        return -abs(f ^ all)
+
+    tnopt = qtn.TNOptimizer(
+        psi0,
+        loss_fn=loss,
+        shared_tags=[ 'p'+f'{i}' for i in range(2*depth) ],
+        loss_constants={'target': T_gs},
+        autodiff_backend=backend,
+        loss_target=-0.99,
+    )
+    # correct number of opt params identified
+    assert len(tnopt.variables) == 2*depth 
+
+    psi_opt = tnopt.optimize_basinhopping(n=10, nhop=5)
+    assert sum(loss < -0.99 for loss in tnopt.losses) == 1
```

Comment:
By the way would be good to test here a couple of things related to the shared tags functionality...
1. maybe that the TNOptimizer vector space is as expected smaller (i.e. 2 * 2 * depth?).
2. the data injected into ``psi_opt`` actually matches between shared tags tensors

Reply:
Few comments here: I tested it for a periodic MPS, as it worked. I tried different scenarios also (same tensors in the bulk) still works well. 

As a QUIMB user, I think share-tags and tags should be totally different (no common elements), that at the least for me makes more sense as I see what I am targeting. 

Maybe const_tags are redundant here makes things more complicated.

Reply:
The problem I am facing is that results are not consistant. See the following example: periodic MPS with one variable and multiple variable for a periodic Hamiltonian. It seems in periodic MPS with one common tensor, it can not auto-diff correctly as energy is much higher than the same MPS with multiple tensors.   



`
%matplotlib inline
from quimb import *
from quimb.tensor import *
import quimb as qu
import quimb.tensor as qtn
import numpy as np
from orderedset import OrderedSet
L = 6
psi = MPS_rand_state(L, 12, phys_dim=2, normalize=True, cyclic=True, seed=10)
psi.show()
ham = MPO_ham_heis(L, j=(1.0,1.0,1.0), bz=0.0, S=0.5, cyclic=True)

E_exact = qu.heisenberg_energy(L)
print ("energy_exact", L, E_exact)

for i in range(L):
 psi[i].add_tag("I")


print (psi)


def norm_mps(psi):
    norm_val=(psi.H & psi).contract()
    psi=psi*(norm_val**(-0.5))
    return psi

def energy_f(psi, ham):
    p_H=psi.H 
    psi.align_(ham, p_H)
    return ( p_H & ham & psi).contract()


def loss_f(psi, p_DMRG):
    return  1-abs((psi & p_DMRG.H).contract(all, optimize='auto-hq'))

tnopt = qtn.TNOptimizer(
    psi,                         
    loss_fn=energy_f,                       
    norm_fn=norm_mps,                         
    loss_constants={ "ham" : ham },  
    tags={  },                 
    shared_tags={"I"},
    constant_tags={ },                   
    optimizer='adam',             
    autodiff_backend='torch',         
)

psi=tnopt.optimize(500, tol=1e-12, ftol=1e-12, gtol=1e-12)


psi = MPS_rand_state(L, 6, phys_dim=2, normalize=False, cyclic=True)
for i in range(L):
 psi[i].add_tag("I")


tnopt = qtn.TNOptimizer(
    psi,                         
    loss_fn=energy_f,                       
    norm_fn=norm_mps,                         
    loss_constants={ "ham" : ham },  
    tags={ "I" },                 
    shared_tags={},
    constant_tags={ },                   
    optimizer='L-BFGS-B',             
    autodiff_backend='torch',         
)

psi=tnopt.optimize(220, tol=1e-12, ftol=1e-12, gtol=1e-12)

`

Reply:
I tried this with even and odd tensor with shared data and the energy was good - I think something to do with having a large enough unit cell rather than the autodiff implementation no?

Reply:
Ah, smart. The Heisenberg model has an anti-Ferro magnet structure requiring a two-unit cell ...AB...probably that is the reason.

Seems it work well :)